### PR TITLE
sign -- given a certificate signing request, and a CA (key, certificate), sign that request

### PR DESCRIPTION
**Pull request by hannesm**
also:
- reimplements selfsign in terms of sign
- messes around with X509v3 extensions:
- always puts subject key id and authority key id X509v3 extensions
- sign optionally inserts subjectAlternativeName extension
- keyUsage and extendedKeyUsage extensions are set
- sets basicConstraints = true in CA certificates, = false in client / server certificates

### CODE REVIEW

File: src/common.ml [hunk 1]
```diff
@@ -62,7 +120,11 @@ let days =
   let doc = "The number of days from the start date on which the " ^ thing ^ " will expire." in
   Arg.(value & opt int 365 & info ["d"; "days"] ~doc)
 
-let common_name = 
+let is_ca =
+  let doc = "Sign a CA cert (and include appropriate extensions)." in
+  Arg.(value & flag & info ["C"; "is_ca"] ~doc)
```
Comment (yomimono) [inline src/common.ml]: I'd prefer just "ca" rather than "is_ca" for the long version.

File: src/sign.ml [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  | _, Common.Error str, _
+  | _, _, Common.Error str -> Printf.eprintf "%s\n" str; `Error
+
+let client =
+  let doc = "Client or server certificate" in
+  Arg.(value & flag & info ["client"] ~doc)
+
+let altname =
+  let doc = "include a subjectAltName in certificate" in
+  Arg.(value & flag & info ["altname"] ~doc)
```
Comment (yomimono) [inline src/sign.ml:43]: Can you be more specific in the docstring about what SAN will be set, since it's not user-specified?

File: src/sign.ml [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  let doc = "sign a certificate" in
+  let man = [ `S "BUGS";
+              `P "Submit bugs at https://github.com/yomimono/ocaml-certify";] in
+  Term.info "sign" ~doc ~man
+
+let () =
+  match Term.eval (sign_t, info) with
+  | `Help -> exit 0 (* TODO: not clear to me how we generate this case *)
+  | `Version -> exit 0  (* TODO: not clear to me how we generate this case *)
+  | `Error _ -> exit 1
```
Comment (yomimono) [inline src/sign.ml:81]: `Help and`Version should probably exit 1 as well.

**Comment by yomimono** [comment]
Can you add a few lines about `sign` to the README as well?

**Comment by hannesm** [comment]
addressed your comments, and rebased

**Comment by yomimono** [comment]
Excellent, thanks!

File: src/sign.ml (hunk 3 above)
Comment (dbuenzli) [inline src/sign.ml:81]: @yomimono it seems to me that it's usual for tools to return `0` on `--help` or `--version` invocation if these are sucessfull. @hannesm Regarding the `TODO` comment, these things are documented [here](http://erratique.ch/software/cmdliner/doc/Cmdliner.Term.html#TYPEresult), it's the result you get whenever the tool is invoked with `--version` or with `--help`.

Comment (hannesm) [inline src/sign.ml:81]: @dbuenzli thx! I just copy & pasted that piece of code from [here](https://github.com/yomimono/ocaml-certify/blob/primary/src/selfsign.ml#L31-L35)

**Pull request merged**